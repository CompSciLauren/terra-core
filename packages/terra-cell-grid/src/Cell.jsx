import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './Cell.module.scss';
import CellGridUtils from './_CellGridUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Content to be displayed for the row cell
   */
  children: PropTypes.node,
  /**
   * Function callback for the ref of the td.
   */
  refCallback: PropTypes.func,
  /**
   * Function callback for the ref of the td.
   */
  width: PropTypes.shape({
    static: PropTypes.shape({
      value: PropTypes.number,
      unit: PropTypes.string,
    }),
    percentage: PropTypes.number,
    scalar: PropTypes.number,
  }),
};

const defaultProps = {
  children: [],
};

const Cell = ({
  children,
  refCallback,
  width,
  ...customProps
}) => {
  const cellClassNames = cx([
    'cell',
    customProps.className,
  ]);

  return (
    <div
      {...customProps}
      style={CellGridUtils.styleFromWidth(width)} // eslint-disable-line react/forbid-dom-props
      className={cellClassNames}
      ref={refCallback}
    >
      {children}
    </div>
  );
};

Cell.propTypes = propTypes;
Cell.defaultProps = defaultProps;

export default Cell;
