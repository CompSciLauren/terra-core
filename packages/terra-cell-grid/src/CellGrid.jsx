import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './CellGrid.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The children passed to the component
   */
  children: PropTypes.node.isRequired,
  /**
   * Indicates the desired divider styles to apply to a cell grid and its children.
   * One of `'none'`, `'vertical'`, `'horizontal'`, `'both'`.
   */
  dividerStyle: PropTypes.oneOf(['none', 'vertical', 'horizontal', 'both']),
  /**
  /**
   * Function callback for the ref of the cell grid.
   */
  refCallback: PropTypes.func,
};

const defaultProps = {
  dividerStyle: 'none',
};

const CellGrid = ({
  children,
  dividerStyle,
  refCallback,
  ...customProps
}) => {
  const cellGridClassNames = cx([
    `divider-${dividerStyle}`,
    'cell-grid',
    customProps.className,
  ]);

  return (
    <div {...customProps} className={cellGridClassNames} ref={refCallback}>
      {children}
    </div>
  );
};

CellGrid.propTypes = propTypes;
CellGrid.defaultProps = defaultProps;

export default CellGrid;
