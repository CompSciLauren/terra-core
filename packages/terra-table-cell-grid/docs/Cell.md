# Cell

The Cell is a sub component to be used inside a cell grid.

## Usage

```jsx
import React from 'react';
import { Cell } from 'terra-table-cell-grid';

<Cell key="row-0-cell-0">
  Table row text
</Cell>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
