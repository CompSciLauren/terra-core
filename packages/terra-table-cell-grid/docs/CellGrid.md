# Cell Grid

The CellGrid is a sub component that groups table cells passed through children. Each row handles it's own selection state through props. The onSelect function that can be passed to the cell grid will return the metaData associated to the cell grid. The metaData allows for greater control of state with the ability to associate any indicator or function with the cell grid.

## Usage

```jsx
import React from 'react';
import { CellGrid, Cell } from 'terra-table-cell-grid';

  <CellGrid
    key="unique-react-key"
    isSelectable
    isSelected
    metaData={{ myInfo: 'my-data' }}
    onSelect={this.handleRowSelection}
  >
    <Cell>
      Cell Content
    </Cell>
    <Cell>
      Cell Content
    </Cell>
  </CellGrid>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
