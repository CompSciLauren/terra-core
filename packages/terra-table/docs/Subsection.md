# Subsection

The Terra Table Subsection is a sub component that creates a subsection header based on the given props and groups the child cell grids beneath it. The subsection header has the ability to toggle the display of its child rows given the isCollapsible and isCollapsed props. For a subsection to be collapsed both isCollapsible and isCollapsed must be true. Each section handles its collapsible and collapsed state.

## Usage

```jsx
import React from 'react';
import { CellGrid, Subsection } from 'terra-table';

  <Subsection
    key="unique-react-key"
    title="Subsection Title"
    isCollapsible
    isCollapsed
    metaData={{ myInfo: 'my-data' }}
    onSelect={this.handleRowSelection}
  >
    <Row />
    <Row />
    <Row />
  </Subsection>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
