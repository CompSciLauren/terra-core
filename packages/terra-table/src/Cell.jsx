import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './Cell.module.scss';
import { styleFromWidth } from './TableUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Child content to be displayed for the row cell.
   */
  children: PropTypes.node,
  /**
   * Whether or not the cell should adhere to the tables paddingStyle.
   */
  isPadded: PropTypes.bool,
  /**
   * Function callback returning the html node for the cell.
   */
  refCallback: PropTypes.func,
  /**
   * Width of the cell. Should match header cell counter-part.
   */
  width: PropTypes.shape({
    /**
     * Static width that for the cell.
     */
    static: PropTypes.shape({
      /**
       * Numerical width value.
       */
      value: PropTypes.number.isRequired,
      /**
       * Valid css units are supported (i.e. 'px', 'rem', etc).
       */
      unit: PropTypes.string.isRequired,
    }),
    /**
     * Percentage width of the row for the cell.
     */
    percentage: PropTypes.number,
    /**
     * Relative scalar value of the cell's width compared to its sibling cells.
     */
    scalar: PropTypes.number,
  }),
};

const defaultProps = {
  children: [],
  isPadded: false,
};

const Cell = ({
  children,
  isPadded,
  refCallback,
  width,
  ...customProps
}) => {
  const cellClassNames = cx([
    'cell',
  ]);

  let content = children;
  if (isPadded) {
    content = (
      <div className={cx('container')}>
        {content}
      </div>
    );
  }

  return (
    <div
      {...customProps}
      style={styleFromWidth(width)} // eslint-disable-line react/forbid-dom-props
      className={customProps.className ? `${cellClassNames} ${customProps.className}` : cellClassNames}
      ref={refCallback}
      role="gridcell"
    >
      {content}
    </div>
  );
};

Cell.propTypes = propTypes;
Cell.defaultProps = defaultProps;

export default Cell;
