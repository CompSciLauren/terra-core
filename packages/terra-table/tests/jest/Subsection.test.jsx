import React from 'react';
import Subsection from '../../src/Subsection';
import CellGrid from '../../src/CellGrid';
import Cell from '../../src/Cell';

// Snapshot test
it('should render a default subsection', () => {
  const shallowComponent = shallow(
    <Subsection title="test" />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a subsection with a cellgrid', () => {
  const shallowComponent = shallow(
    <Subsection title="test">
      <CellGrid>
        <Cell />
      </CellGrid>
    </Subsection>,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a subsection with isCollpased', () => {
  const shallowComponent = shallow(
    <Subsection title="test" isCollapsed />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a subsection with isCollapsible', () => {
  const shallowComponent = shallow(
    <Subsection title="test" isCollapsible />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a subsection with level', () => {
  const shallowComponent = shallow(
    <Subsection title="test" level={4} />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a subsection with metaData', () => {
  const shallowComponent = shallow(
    <Subsection title="test" metaData={{}} />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a subsection with onSelect', () => {
  const shallowComponent = shallow(
    <Subsection title="test" onSelect={jest.fn()} />,
  );
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a subsection with refCallback', () => {
  const shallowComponent = shallow(
    <Subsection title="test" refCallback={jest.fn()} />,
  );
  expect(shallowComponent).toMatchSnapshot();
});
