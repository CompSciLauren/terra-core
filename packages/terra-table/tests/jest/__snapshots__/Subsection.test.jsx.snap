// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a default subsection 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with a row 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    title="test"
  />
  <Row
    dividerStyle="none"
    isDisabled={false}
    isSelectable={false}
    isSelected={false}
    isStriped={false}
    selectionStyle="default"
  >
    <Cell />
  </Row>
</Fragment>
`;

exports[`should render a subsection with isCollapsible 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={true}
    level={3}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with isCollpased 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={true}
    isCollapsible={false}
    level={3}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with level 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={4}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with metaData 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    metaData={Object {}}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with onSelect 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    onSelect={[MockFunction]}
    title="test"
  />
</Fragment>
`;

exports[`should render a subsection with refCallback 1`] = `
<Fragment>
  <SubsectionHeader
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    refCallback={[MockFunction]}
    title="test"
  />
</Fragment>
`;
